From 00d3c8ccdd839387bbbdcabfbddd2d9415a87cb5 Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Mon, 13 Mar 2017 12:35:24 +0200
Subject: [PATCH 1561/5242] ARM: cpuidle-imx*: Report the entered state index

commit  3e4326a608cfc6f7397f91205b5b4bfc9b2610e2 from
https://source.codeaurora.org/external/imx/linux-imx.git

When a cpuidle driver is called it can choose to enter a different state
that what was asked from above. When this happens it should return the
actual entered_state index for proper accounting.

This fixes the various imx cpuidle drivers which depend on low busfreq
for LOW-POWER-IDLE to correctly report that they entered WAIT instead.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/cpuidle-imx6sll.c |    1 +
 arch/arm/mach-imx/cpuidle-imx6sx.c  |    1 +
 arch/arm/mach-imx/cpuidle-imx6ul.c  |    1 +
 arch/arm/mach-imx/cpuidle-imx7d.c   |    1 +
 4 files changed, 4 insertions(+)

diff --git a/arch/arm/mach-imx/cpuidle-imx6sll.c b/arch/arm/mach-imx/cpuidle-imx6sll.c
index 8e06f33..65d82e5 100644
--- a/arch/arm/mach-imx/cpuidle-imx6sll.c
+++ b/arch/arm/mach-imx/cpuidle-imx6sll.c
@@ -97,6 +97,7 @@ static int imx6sll_enter_wait(struct cpuidle_device *dev,
 
 	imx6_set_lpm(WAIT_UNCLOCKED);
 	if ((index == 1) || ((mode != BUS_FREQ_LOW) && index == 2)) {
+		index = 1;
 		cpu_do_idle();
 	} else {
 		imx_gpc_switch_pupscr_clk(true);
diff --git a/arch/arm/mach-imx/cpuidle-imx6sx.c b/arch/arm/mach-imx/cpuidle-imx6sx.c
index c1c7ad4..be73709 100644
--- a/arch/arm/mach-imx/cpuidle-imx6sx.c
+++ b/arch/arm/mach-imx/cpuidle-imx6sx.c
@@ -109,6 +109,7 @@ static int imx6sx_enter_wait(struct cpuidle_device *dev,
 
 	imx6_set_lpm(WAIT_UNCLOCKED);
 	if ((index == 1) || ((mode != BUS_FREQ_LOW) && index == 2)) {
+		index = 1;
 		cpu_do_idle();
 	} else {
 			/* Need to notify there is a cpu pm operation. */
diff --git a/arch/arm/mach-imx/cpuidle-imx6ul.c b/arch/arm/mach-imx/cpuidle-imx6ul.c
index 8e0fdf7..7708d87 100644
--- a/arch/arm/mach-imx/cpuidle-imx6ul.c
+++ b/arch/arm/mach-imx/cpuidle-imx6ul.c
@@ -98,6 +98,7 @@ static int imx6ul_enter_wait(struct cpuidle_device *dev,
 	imx6_set_lpm(WAIT_UNCLOCKED);
 	if ((index == 1) || ((mode != BUS_FREQ_LOW) && index == 2)) {
 		cpu_do_idle();
+		index = 1;
 	} else {
 		/*
 		 * i.MX6UL TO1.0 ARM power up uses IPG/2048 as clock source,
diff --git a/arch/arm/mach-imx/cpuidle-imx7d.c b/arch/arm/mach-imx/cpuidle-imx7d.c
index 4ba99d8..83d1f48 100644
--- a/arch/arm/mach-imx/cpuidle-imx7d.c
+++ b/arch/arm/mach-imx/cpuidle-imx7d.c
@@ -94,6 +94,7 @@ static int imx7d_enter_low_power_idle(struct cpuidle_device *dev,
 	int mode = get_bus_freq_mode();
 
 	if ((index == 1) || ((mode != BUS_FREQ_LOW) && index == 2)) {
+		index = 1;
 		if (atomic_inc_return(&master_wait) == num_online_cpus())
 			imx_gpcv2_set_lpm_mode(WAIT_UNCLOCKED);
 
-- 
1.7.9.5

